book/src/contributing/ast_ebnf.md

```diff
@@ -0,0 +1,237 @@
+We currently take inputs from the following AST (visualization can be done
+using <https://rr.red-dove.com/ui>). Literals are strings, numbers and
+booleans
+
+``` ebnf
+char ::= [a-zA-Z]
+string ::= char*
+digit ::= [0-9]
+uint ::= digit+
+int ::= ("-")? uint
+float ::= int (".")? uint
+bool ::= "true" | "false"
+
+literal ::=
+| "\"" string "\""
+| "'" char "'"
+| int
+| float
+| bool
+```
+
+We support a number of simple types characters, strings, booleans and
+numbers. Number types for integers (8,16,32,64,128 bit or machine sized)
+and floats (16,32, or 64 bit). Composite types are tuples, fixed length
+lists (arrays), variable length lists (vectors/slices), ptr types, and
+function types. Lastly we have named types defined by items, e.g. enums
+and structs.
+
+``` ebnf
+generic_value ::=
+| "'" ident
+| ty
+| expr
+
+impl ::=
+| "self"
+| goal
+| "dyn"
+| <!TODO!>
+goal ::= ident "<" (generic_value ",")* ">"
+
+ty ::=
+| "bool"
+| "char"
+| "u8" | "u16" | "u32" | "u64" | "u128" | "usize"
+| "i8" | "i16" | "i32" | "i64" | "i128" | "isize"
+| "f16" | "f32" | "f64"
+| "str"
+| (ty ",")*
+| "[" ty ";" int "]"
+| "[" ty "]"
+| "*const" ty | "*mut" ty
+| "*" expr | "*mut" expr
+| ident
+| (ty "->")* ty
+| impl "::" ident
+| "impl" ty
+| (goal)*
```

Comment:
I don't get this (goal)* line here. From what variant in AST does this come from?